pkg/cloud/packet/scope/machine.go

```diff
@@ -208,3 +237,160 @@ func (m *MachineScope) GetRawBootstrapData() ([]byte, error) {
 
 	return value, nil
 }
+
+func getProviderIDPrefix(ctx context.Context, mgmtClient client.Client, workloadClientGetter remote.ClusterClientGetter, cluster *clusterv1.Cluster, machine *clusterv1.Machine, packetMachine *infrav1.PacketMachine) (string, error) {
+	// Use existing prefix if already defined
+	if existingPrefix := providerIDPrefixFromPacketMachine(packetMachine); existingPrefix != "" {
+		return existingPrefix, nil
+	}
+
+	// Try to determine the appropriate prefix from any known cloud-provider deployments
+	fromDeployments, err := providerIDFromCloudProviderDeployments(ctx, mgmtClient, workloadClientGetter, cluster)
+	if err != nil {
+		return "", err
+	}
+	if fromDeployments != "" {
+		return fromDeployments, nil
+	}
+
+	// TODO: maybe switch to unstructured?
+	// TODO: add note that this only works for kubeadm bootstrapping, might need
+	// similar support for alternative bootstrappers
+	if machine.Spec.Bootstrap.ConfigRef != nil && machine.Spec.Bootstrap.ConfigRef.Kind == "KubeadmConfig" {
+		kubeadmConfig := new(bootstrapv1.KubeadmConfig)
+		key := client.ObjectKey{
+			Name:      machine.Spec.Bootstrap.ConfigRef.Name,
+			Namespace: packetMachine.Namespace,
+		}
+		// TODO: does this need to take into account notfound?
+		if err := mgmtClient.Get(ctx, key, kubeadmConfig); err != nil {
+			return "", fmt.Errorf("failed to get bootstrap resource: %w", err)
+		}
+
+		// providerid being set explicitly in the InitConfiguration
+		if kubeadmConfig.Spec.InitConfiguration != nil && kubeadmConfig.Spec.InitConfiguration.NodeRegistration.KubeletExtraArgs != nil {
+			if value, ok := kubeadmConfig.Spec.InitConfiguration.NodeRegistration.KubeletExtraArgs["provider-id"]; ok {
+				if parsed, err := noderefutil.NewProviderID(value); err == nil {
+					return parsed.CloudProvider(), nil
+				}
+			}
+		}
+
+		// providerid being set explicitly in the JoinConfiguration
+		if kubeadmConfig.Spec.JoinConfiguration != nil && kubeadmConfig.Spec.JoinConfiguration.NodeRegistration.KubeletExtraArgs != nil {
+			if value, ok := kubeadmConfig.Spec.JoinConfiguration.NodeRegistration.KubeletExtraArgs["provider-id"]; ok {
+				if parsed, err := noderefutil.NewProviderID(value); err == nil {
+					return parsed.CloudProvider(), nil
+				}
+			}
+		}
+
+		// inspect postkubeadmcommands
+		for i := range kubeadmConfig.Spec.PostKubeadmCommands {
+			if strings.Contains(kubeadmConfig.Spec.PostKubeadmCommands[i], "github.com/packethost/packet-ccm") {
+				return deprecatedProviderIDPrefix, nil
+			}
+
+			if strings.Contains(kubeadmConfig.Spec.PostKubeadmCommands[i], "github.com/equinix/cloud-provider-equinix-metal") {
+				return providerIDPrefix, nil
+			}
+		}
+	}
+
+	return providerIDPrefix, nil
+}
+
```

Comment:
I could use an explanation comment of how and why providerIDFromCloudProviderDeployments is expected to work.

Reply:
Added some additional comments around the rationale and expected behavior of the functions